internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +59,682 @@ type VolumeGroupReplicationReconciler struct {
+*/
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 func (r *VolumeGroupReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx, "Request.Name", req.Name, "Request.Namespace", req.Namespace)
+
+	// Fetch VolumeGroupReplication instance
+	instance := &replicationv1alpha1.VolumeGroupReplication{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("volumeGroupReplication resource not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Get VolumeGroupReplicationClass instance
+	vgrClassObj, err := r.getVolumeGroupReplicationClass(logger, instance.Spec.VolumeGroupReplicationClassName)
+	if err != nil {
+		logger.Error(err, "failed to get volumeGroupReplicationClass resource", "VGRClassName", instance.Spec.VolumeGroupReplicationClassName)
+		_ = r.setGroupReplicationFailure(instance, logger, err)
+		return reconcile.Result{}, err
+	}
+
+	// Validate that required parameters are present in the VGRClass resource
+	err = validatePrefixedParameters(vgrClassObj.Spec.Parameters)
+	if err != nil {
+		logger.Error(err, "failed to validate parameters of volumeGroupReplicationClass", "VGRClassName", instance.Spec.VolumeGroupReplicationClassName)
+		_ = r.setGroupReplicationFailure(instance, logger, err)
+		return reconcile.Result{}, err
+	}
+
+	// Declare all dependent resources
+	vgrContentName := fmt.Sprintf("vgrcontent-%s", instance.UID)
+	if instance.Spec.VolumeGroupReplicationContentName != "" {
+		vgrContentName = instance.Spec.VolumeGroupReplicationContentName
+	}
+	vgrContentObj := &replicationv1alpha1.VolumeGroupReplicationContent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: vgrContentName,
+		},
+	}
+	vrName := fmt.Sprintf("vr-%s", instance.UID)
+	if instance.Spec.VolumeReplicationName != "" {
+		vrName = instance.Spec.VolumeReplicationName
+	}
+	vrObj := &replicationv1alpha1.VolumeReplication{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      vrName,
+			Namespace: instance.Namespace,
+		},
+	}
+
+	// Create/Update dependent resources only if the instance is not marked for deletion
+	if instance.GetDeletionTimestamp().IsZero() {
+
+		// Add finalizer to VGR instance
+		if err = AddFinalizerToVGR(r.Client, logger, instance); err != nil {
+			logger.Error(err, "failed to add VolumeGroupReplication finalizer")
+			return reconcile.Result{}, err
+		}
+
+		// Check if PVCs exist based on provided selectors
+		pvcList, pvHandlesList, labelSelector, err := r.getMatchingPVCsFromSource(instance, logger, vgrClassObj)
+		if err != nil {
+			logger.Error(err, "failed to get PVCs using selector")
+			_ = r.setGroupReplicationFailure(instance, logger, err)
+			return reconcile.Result{}, err
+		}
+		if len(pvcList) > 100 {
```

Comment:
Let's make this configurable with default as 100  ? 
similar to https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/csi-addons-config.md

Reply:
Sounds good, but probably we can take it up as a follow up PR?

Reply:
It is fine to make this change in a follow-up PR. Please open an issue and link it here.

Reply:
Created issue #787 for it